### ISSUE TITLE
Refactor the whole Spring Security configuration to use OAuth2 only

### ISSUE THREAD

**Issue by jdubois** [jhipster/generator-jhipster#2584]
I have 2 issues with our current security options:

- we support 4 types of authentication, and this making things really complex as we have too many use cases. This is why we have too many opened issues here.
- we don't provide anything for working with "microservices"

I've done some research here, and I'm clearly not an expert in that field, but I think we should only support Spring Security's OAuth2 mechanism:

- It has a JWT token support, so it could be quite light and very close to our current XAuth support (which wouldn't be useful anymore)
- It should be able to be either generated in our current application (this is the current "monolithic" approach) or use an external Spring Security OAuth2 server (that would be the "microservices" approach)

Now there are a few things I don't know about this:

- Can it work with our "social login" option? We need to keep those, of course.
- Can we make this work with the "remember me" filter? This is an important feature to me.
- Looks like the CSRF token is still needed -> this isn't an issue as we already have this working, but I thought CSRF wasn't an issue with OAuth2
- Can we still have our current "session invalidation service" in the admin menu? I find it really cool.

Any comments, ideas?

**Comment by deepu105** [comment]
does session invalidation work?? last I tried it didnt work and i forgot to follow that up, what happens if i invalidate my current active session?

**Comment by jdubois** [comment]
Session invalidation only removes the current "remember me" tokens -> so for the current OAuth2 implementation (which doesn't have remember me), it shouldn't work.
But I need to check if this could be implemented.

**Comment by cbornet** [comment]
For what I know about OAuth2:
* As there is no session/server cookie involved, CSRF is not needed.
* The "remember-me" can be a refresh token stored in local storage or client cookie
* To invalidate a "session", you need to revoke both the refresh token and the access token. Spring security has a [revokeToken](https://github.com/spring-projects/spring-security-oauth/blob/234ec98f88f9ef5246381beed21eb7345d4aa82d/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/DefaultTokenServices.java#L267) method which seems to do the job.

**Comment by cbornet** [comment]
@jdubois what is your source stating that CSRF protection is needed ? There are some "anti-CSRF" protection mechanisms involved with authorization/implicit flows but they come out-of-the box with Spring Security OAuth

**Comment by jdubois** [comment]
@cbornet I'm reading this here: http://projects.spring.io/spring-security-oauth/docs/oauth2.html

**Comment by cbornet** [comment]
This is only for /oauth/confirm_access and /oauth/error since they are forms used in the authorization_code/implicit flows. For jhipster login, we use the password flow so these endpoints are not used.

**Comment by jdubois** [comment]
So for you there is no CSRF filter to add? That would be a lot better!

**Comment by sdoxsee** [comment]
If you're going with OAuth2 (which I think is a great idea), I'd recommend proxying all of your API calls via a thin server side component or API gateway (example here: https://spring.io/guides/tutorials/spring-security-and-angular-js/#_sso_with_oauth2_angular_js_and_spring_security_part_v). That way you'll use the (more secure) "authorization code" flow rather than the (problematic) "implicit" flow (see http://alexbilbie.com/2014/11/oauth-and-javascript/).

I wouldn't recommend that jhipster keep using the password flow (at least not directly from the SPA). The password flow isn't even enabled "out-of-the-box" with spring security oauth2.

Using a server-side proxy still requires CSRF but I think that's a good thing...safer.

**Comment by jdubois** [comment]
@sdoxsee yes that's the idea, to have a Zuul gateway in front

For the password flow: we still need to have one, we can't remove this

**Comment by cbornet** [comment]
I know the Alex Bilbie blog post and I have still not understood his point:
* if you don't store the access_token, another app can't get your access_token
* if you store it, you need physical access to the computer to get the token but that's the same security issue than with a remember-me. CSRF attacks won't work because it's a client cookie, not a server one.
* if you store it encrypted someone can still use it encrypted and simply use your proxy...

**Comment by cbornet** [comment]
BTW I would be very grateful if someone pointed out something I missed !

**Comment by sdoxsee** [comment]
@jdubois why do we still need to use the password flow? 

Authentication is an Authorization Server (or better...an OpenID Connect Provider) job and the collection of credentials should be done there (IMO).

@cbornet 
* if you don't store the access_token, another app can't get your access_token

If we're using an API gateway, the tokens are safe on the server.
* if you store it, you need physical access to the computer to get the token but that's the same security issue than with a remember-me. CSRF attacks won't work because it's a client cookie, not a server one.

by "client-side cookie" do you mean HTML 5 Web Storage like sessionStorage and localStorage? If so, XSS is a possiblity (https://stormpath.com/blog/where-to-store-your-jwts-cookies-vs-html5-web-storage/)
* if you store it encrypted someone can still use it encrypted and simply use your proxy...

Again, you probably don't want your tokens in your browser. That's the advantage of an API gateway or server-side proxy that handles tokens for you.

**Comment by cbornet** [comment]
@sdoxsee No I mean a cookie field written by your client javascript. It's much like localStorage and XSS is also a possibility there. The libraries we use should prevent XSS but of course you never know...

**Comment by sdoxsee** [comment]
@jdubois I also just found this from Dave Syer regarding a login form on the gateway (which must be the password flow). https://github.com/spring-guides/tut-spring-security-and-angular-js/issues/57#issuecomment-167597115

"The whole point of the auth server is that it is a centralized resource, so putting a login form in the gateway (unless it is also the auth server) defeats the object somewhat. I don't think what you are asking makes a lot of sense consequently, and it would require custom secure code in the client which I think is a bad idea"

**Comment by cbornet** [comment]
As for using authorization_code flow this is poor user experience since you first login as a user then need to authorize the app. Why would you have to authorize your own app ?

**Comment by sdoxsee** [comment]
Hi @cbornet. I agree that having to "authorize" your own app every time (or even once for some clients) would make for a bad user experience. However, the authorization server can easily be configured to "whitelist" or "autoApprove" scopes for specific clients so that the jhipster webapp user would never have to manually "approve" anything on the auth server (e.g. https://github.com/spring-projects/spring-security-oauth/blob/master/samples/oauth2/sparklr/src/main/java/org/springframework/security/oauth/examples/sparklr/config/OAuth2ServerConfig.java#L156)

**Comment by sdoxsee** [comment]
I've modified the spring tutorial example with the UI gateway, authorization server, and resource server using JWTs. Note, I'm not using the resource server in this example...so you only need to start the authserver and ui spring-boot apps.

https://github.com/sdoxsee/jhipster-with-external-authserver

The UI (jhipster) project only really has jhipster angular code with the vanilla server-side parts being provided by Spring defaults. It's using Zuul to proxy to the authorization server to look up the user account at '/user'. I've autoApproved the jhipster (or in this case "acme") client so that there is no consent screen. The only credentials that are setup on the authserver are "user/password".

We can customize the authorization server login screen to look better and I still need to add an end_session_endpoint (à la OpenID Connect http://openid.net/specs/openid-connect-session-1_0-17.html) to the authorization server to clear the session on the authorization server when the client logs out (which may or may not be desired). For now, to logout of the authorization server, you'll need to manually go to localhost:9999/uaa/login and clear the cookies from the browser or the jhipster client will log in automatically without being prompted for credentials again since the a jhipster logout only logs out of the jhipster client.

I'd need to spend more time with it to make sure it's doing the right things but it illustrates the authorization code flow with a server-side proxy keeping the tokens and client secrets away from the browser.

**Comment by cbornet** [comment]
@sdoxsee Even with authorization whitelisting, I don't see what would be the added value compared to password flow apart that you would use the same login form for your app and client apps : you still have the client_id/client_secret exposed in your JS and if you want a "remember-me", you still have to store the refresh_token in localStorage or client cookie.
As for using a server-side proxy, this is certainly added value in a micro-service architecture as this would reduce the attack surface regarding XSS but I don't think we can require it for the standalone jhipster.
@jdubois I checked the current code and it seems that both the access_token and refresh_token are stored in localStorage after login. So we have remember-me activated by default...  I think this should only be done if the user checks a "remember-me" box. And the JS code doesn't event use the refresh token (which has a validity of 30 days).

**Comment by cbornet** [comment]
@jdubois To go back to the initial question and try to sum up , if you want to use OAuth2 with "remember-me" in the "monolithic" JHipster then you are sensitive to XSS attacks on your localStorage.

**Comment by jdubois** [comment]
Thanks @sdoxsee @cbornet for all the feedback.

In fact I'm now thinking this ticket was a mistake, I will not do this refactoring. Let me summarize, and get your comments before closing this:

- Currently our 3 authentication schemes work differently (for example for remember-me): my idea was to clean that up, but in the end this isn't really a problem. They are just different use-cases.
- OAuth2 with SSO is bad from a UI perspective: you need to be redirected to an OAuth2 server (breaking the SPA)
- With OAuth2 the remember-me can't work nicely (so either you have XSS attacks, or your users are logged out after a certain time, either way it's worse that our current default mechanism).
- OAuth2 is good if you want to do SSO: but for what I want to do in our microservices approach, this isn't such a good use-case. It could be an "advanced use-case", don't get me wrong, but by default I'd rather just have the gateway validate all the requests, and then have no security on the back-end microservices. It's much easier to setup, and performance will be much, much better (another issuer I have with this OAuth2 - SSO approach is that we would need a top-notch OAuth2 server, and I haven't found anything guaranteeing that Spring Security OAuth2 or one of its competitors could handle that load and high-availability)

**Comment by cbornet** [comment]
+1

**Comment by sdoxsee** [comment]
Thanks for the great discussion @jdubois and @cbornet 

@cbornet with the gateway solution and authorization_code flow, neither the tokens nor the client secret are exposed to the JS. 

@jdubois you bring up some interesting points
> OAuth2 with SSO is bad from a UI perspective: you need to be redirected to an OAuth2 server (breaking the SPA)

I agree that this is a drawback. However, we could save state in cookies or HTML 5 web storage to remember where we left off and maybe have the rememberme on the auth server. Alternatively, we could continue with password flow but make the token request from the Gateway (server-side) to keep the client secret and tokens out of the browser? It might even work with rememberme on the webapp itself because we're authenticating with the authorization server server-side rather than through the browser.

> With OAuth2 the remember-me can't work nicely (so either you have XSS attacks, or your users are logged out after a certain time, either way it's worse that our current default mechanism).

I'm not sure that's true. If the client is the SPA then I agree. If the client is the webapp with an SPA front-end then I think we'd get rememberme + safety from XSS attacks. Of course, I'd need to try this out to be sure.

> OAuth2 is good if you want to do SSO: but for what I want to do in our microservices approach, this isn't such a good use-case. It could be an "advanced use-case", don't get me wrong, but by default I'd rather just have the gateway validate all the requests, and then have no security on the back-end microservices. It's much easier to setup, and performance will be much, much better (another issuer I have with this OAuth2 - SSO approach is that we would need a top-notch OAuth2 server, and I haven't found anything guaranteeing that Spring Security OAuth2 or one of its competitors could handle that load and high-availability)

I would argue that it's exactly what we want to do for microservices. If there's no security on the back-end microservices, then are we really building microservices or a bigger monolith with more pieces and complexity? Why go microservices if not to let these services stand independently and be used for any client application that wants integration access and is authorized to do so? 

I'm not sure that performance would be hurt by securing the microservices. All they are doing is receiving self-contained JWT access tokens and verifying them locally. see here: http://nordicapis.com/how-to-control-user-identity-within-microservices/

As for a top-noth OAuth2 server, I don't think the auth server would be hit much at all (relatively). The auth server's job is to authenticate a user and give back tokens. After that it's job is done until the token expires and it needs to be refreshed. Having an authserver is the microservices way of doing things. It has a very specific job that it does very well.

I think that oauth2 is the way to handle microservices and would be happy to discuss it more (by voice or whatever) if that would help. My twitter handle is @doxsees if you want to get in touch. 

Regardless of which way we go, I'm enjoying using jhipster for several projects and appreciate the thought and work that has gone into it.

**Comment by jdubois** [comment]
Great points @sdoxsee especially the last one:

- I've checked with several companies, and from my understanding they just need security at the Gateway level. I'm guessing that it's what Netflix means by an "edge server": behind this server, you are in your own corporate network, and you are secured.
- On the other hand, I agree with you that having SSO looks better in terms of security and de-coupling. Indeed, the OAuth2 server performance shouldn't be an issue if we use JWT tokens, and they would also help to carry out more information than in my first point.

So, basically we have 2 architectures here: I'd like first to make the 1st one work well, as it's simpler, before digging into the 2nd one.

**Comment by jdubois** [comment]
I'm closing this and doing a new ticket at https://github.com/jhipster/jhipster-experimental-microservices (which will be linked here once it's created), as in the end this is a discussion only on microservices.

**Comment by cbornet** [comment]
If we talk about OAuth2 servers, maybe [Tyk](https://tyk.io/) and [WSO2](http://wso2.com/) are worth mentionning since they are pretty popular. Anyway, I think we should propose a reference implementation with one of the OAuth server but a user should be able to easily switch to another one. I don't know if it is possible though...

**Comment by nivemaham** [comment]
Hi @jdubois , @cbornet , I have been following the issues related to using `refresh token` with JWT. I see there has been a PR to do this for `microservices` set-up. I have a jhipster **monolithic** application that uses **`OAuth2` and `JWT`.** Now we have a requirement to prolong the session by "auto-refreshing" the access-token using refresh-token. I'm not sure whether there has been any development done on this? Would you be able to give me some directions, inputs , comments on this?
Would it be possible to do this on a monolithic setup at all using standard jhipster generation? or if i have to do it manually, is there any directions that i could look into?

**Comment by jdubois** [comment]
@nivemaham this is a pretty old issue (for JHipster!), so it's probably not very relevant anymore -> we have added OpenID Connect support recently, which should do what you are looking for. This support is very new (latest release!), so be careful with it, but that's where everyone is heading to. Could you have a look at it? If you have questions on it, could you use Stackoverflow? I'm guessing you won't be the only one!
